### DISCUSSION TITLE
Bumping to 1.1.3 requires extra attention

### DISCUSSION THREAD

**Discussion by TomerFi** [sormuras/junit-platform-maven-plugin#78] [General]
Hey @sormuras 

Just bumped into something weird and figured I'll let you know.
It's not really an issue, if anything, I think it's an improvement.

After bumping from `1.1.2` to `1.1.3` in https://github.com/TomerFi/hebcal-api/commit/1637df76b0090c49fc63b1d475f10a070afc7273.
Some of my test cases started failing with the standard:

```shell
java.lang.IllegalAccessError: class A (in module b) cannot access class C (in module d) because module b does not read module d
```

Not a big deal, I've added the required `reads` permissions in https://github.com/TomerFi/hebcal-api/commit/c7f7efd058b5a0f50d1c9023de1e44a084a6fdd3 and that was it.
It actually makes more sense now, I think...

Nevertheless,
I figured I'll let you know,
since downgrading back to `1.1.2` solved this for me as well.
:-)

**Comment by sormuras** [comment]
Thanks for the heads-up, Tomer.

Looking briefly at `hebcal-api`'s source file layout, I think too that those `reads` directives were missing in the `module-info.test` configuration file.
Yet, that side-effect wasn't expected... indicating a lack of (integration) tests, especially regarding `module-info.test` support.

**Reply by TomerFi** [reply]
Definitely, those `reads` should have been present prior to the version bump.

**Reply by TomerFi** [reply]
@sormuras I bumped into another breaking change with `1.1.3`,
This time it's not a modular project.

[Workflow run](https://github.com/TomerFi/alexa-skill-shabbat-times/runs/3254961320)

**Reply by TomerFi** [reply]
I was just able to work around this by adjusting the dependency-in-question's scope from `provided to compile`.

```xml
<dependency>
  <groupId>com.google.auto.value</groupId>
  <artifactId>auto-value-annotations</artifactId>
  <!-- <scope>provided</scope> -->
</dependency>
```

Any ideas or leads on what might awaken this with `1.1.3`?

**Reply by sormuras** [reply]
That's the commit that changed one line in question:

https://github.com/sormuras/junit-platform-maven-plugin/commit/0cd267107097774b52f019bc5a9da80ce34a12d1

@zregvart suggested it and provided an integration test.

Your use-case was apparently not covered by the original set of tests...

**Reply by zregvart** [reply]
I could be missing some details here, but to me proper scope for `auto-value-annotations` should be `compile`, it's referenced from source like here:

https://github.com/TomerFi/alexa-skill-shabbat-times/blob/deb6e3bf2968ce9c23d7f1a245204be2575dd11c/lambda/src/main/java/info/tomfi/alexa/shabbattimes/Country.java#L22

So the compiler needs to access it.

Bearing in mind that the `@AutoValue` annotation is placed in the class file [(`@Retention(RetentionPolicy.CLASS)`](https://github.com/google/auto/blob/ebfa3ff6911e8a73d86d1dbce440dd4696c5164d/value/src/main/java/com/google/auto/value/AutoValue.java#L44) even more so makes me think that.

I would suggest using [dependency-scope-maven-plugin](https://github.com/HubSpot/dependency-scope-maven-plugin) to assert proper scope usage.

**Reply by TomerFi** [reply]
I see your point,
But I believe something is being overlooked here,
maybe by me.

I use [auto-value](https://github.com/google/auto/blob/master/value/userguide/index.md) for generating source files.
As such, I need it for compile-time only, I have no usage for it at runtime,
nor do I need it as a transitive dependency for my dependants.

As per [maven's docs](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#dependency-scope) the most suitable scope for my needs seems to be `provided`:

```
This is much like compile, but indicates you expect the JDK or a container to provide the dependency at runtime.
...
 A dependency with this scope is added to the classpath used for compilation and test,
but not the runtime classpath. It is not transitive.
```

I've checked using your recommended [dependency-scope-maven-plugin](https://github.com/HubSpot/dependency-scope-maven-plugin),
**no issues** were discovered while using the `provided` scope.

Although I must say, I'm not sure this plugin is related to this particular usage.
As per the [plugin docs](https://github.com/HubSpot/dependency-scope-maven-plugin), if I understand correctly,
it only verifies `test-scoped` dependencies being required by `non-test-scoped` dependencies:

```
This plugin attempts to mitigate this problem by verifying that test-scoped dependencies don't appear elsewhere
in your dependency tree at scope runtime or compile. This allows you to use <scope>test</scope> where it makes
sense and not worry about causing runtime issues.
```

**Reply by TomerFi** [reply]
May I ask what was the desired outcome,
And whether or not it was achieved with https://github.com/sormuras/junit-platform-maven-plugin/commit/0cd267107097774b52f019bc5a9da80ce34a12d1 ?

**Reply by sormuras** [reply]
This commit fixed the issue described in #75

**Reply by TomerFi** [reply]
But...
Isn't that the whole purpose of the `runtime` scope is to exclude dependencies on everything else but runtime,
which also includes testing?

This is based on my understanding of [maven's docs](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#dependency-scope).

**Reply by TomerFi** [reply]
I'll correct myself:

- Dependencies scoped with `runtime`, are included in `runtime` and `testing` classpaths but not in `compile`.
- Dependencies scoped with `provided`, are included in `compile` and `testing` classpaths, but not in `runtime`.

@sormuras 
I think this means that we are all right.
This plugin needs to include dependencies from both the `runtime` and `compile` scopes.

**Reply by TomerFi** [reply]
Maybe [this](https://maven.apache.org/ref/3.8.1/apidocs/org/apache/maven/project/MavenProject.html#getTestClasspathElements--) will do the trick.

**Reply by TomerFi** [reply]
Ok, I pr'd it https://github.com/sormuras/junit-platform-maven-plugin/pull/80.

I've tested it against my configuration and it works well,
Although it broke one integration test (I'm looking into it).

@zregvart can you please verify using `getTestClasspathElements`  works for you as well?

**Reply by zregvart** [reply]
@TomerFi I would expect that `getTestClasspathElements` would work for my use case as it should be a super set of `getRuntimeClasspathElements`, though I think that would break isolation intended with the layered classloading. That is, the _main_ ClassLoader should not contain test classes. I've looked at the documentation of the AutoValue, and there it's [specifically stated](https://github.com/google/auto/blob/ebfa3ff6911e8a73d86d1dbce440dd4696c5164d/value/userguide/index.md#with-maven) that the `com.google.auto.value:auto-value-annotations` should be placed in `compile` scope.
Not sure if it makes sense for you to make use of [`additional*Elements/Dependencies`](https://github.com/sormuras/junit-platform-maven-plugin#plugin-configuration-tweaks) to fine tune the classpath for your use case.

**Reply by TomerFi** [reply]
> I've looked at the documentation of the AutoValue, and there it's [specifically stated](https://github.com/google/auto/blob/ebfa3ff6911e8a73d86d1dbce440dd4696c5164d/value/userguide/index.md#with-maven) that the `com.google.auto.value:auto-value-annotations` should be placed in `compile` scope.

@zregvart  The link you provided states:

```
You will need auto-value-annotations-${auto-value.version}.jar in your compile-time classpath...
```
That doesn't mean you need to set maven's dependency scope to `compile`.
Both the `compile` and `provided` scopes will place the dependency in the compile-time classpath.

More than that, the next paragraph in the link specifically states:

```
... you can set the scope to provided,
so that the user of your library does not have auto-value-annotations as a transitive dependency...
```

(They even have an example to back this up, using the `provided` scope.)

----------------------------------------

I can understand why you're saying `getTestClasspathElements` would break the isolation layer,
and as seen in my pr, you are correct, it does.

But I can't see any reason for the `auto-value-annotations` dependency to be configured with anything else but the `provided` scope for my specific use case.

And I can't say I agree with tweaking the [additional*Elements/Dependencies](https://github.com/sormuras/junit-platform-maven-plugin#plugin-configuration-tweaks) sections.
Code generators using annotation processing, usually are not needed at runtime and are quite common these days,
common enough for this plugin to support it out-of-the-box and not require tweaking.

I guess my next question would be:

Is there a specific reason not to include the compile-time dependencies?
Is there anything wrong with adding both the compile and runtime dependencies?

**Reply by zregvart** [reply]
@TomerFi you are correct and I was mistaking, the `provided` scope is valid scope for `auto-value-annotations`, I should have read the document I pointed to a bit more carefully.
I had a second look at the [workflow run](https://github.com/TomerFi/alexa-skill-shabbat-times/runs/3254961320) you pointed to earlier and I do think that `provided` for your use case is not valid. The [test class](https://github.com/TomerFi/alexa-skill-shabbat-times/blob/deb6e3bf2968ce9c23d7f1a245204be2575dd11c/lambda/src/test/java/info/tomfi/alexa/shabbattimes/ProjectArchTest.java#L148) depends on `com.google.auto.value.AutoValue` so `auto-value-annotations` need to be available in _test_ classloader. This is not something that Maven supports, you can't have the same dependency in two scopes, and it seems that your desire is to have `auto-value-annotations` in `provided` and `test` scopes.

> Is there a specific reason not to include the compile-time dependencies?

I guess answer to this question depends on the on the _where_. In the _main_ classloader? In the _main_ classloader you would have tests running with the classpath that would be different from when they're run in production. To me this looks exactly as the situation what you're trying to avoid: you don't want `auto-value-annotations` on production classpath, or in Maven terms in `compile` or `runtime` scopes; so having the production code run in test with a classpath that will be different from when run in production is very wrong in my opinion.

> Is there anything wrong with adding both the compile and runtime dependencies?

In my opinion, this is what classloader isolation is trying to avoid. Reiterating above, the _main_ classloader is trying to run the production code with classpath as it would be run in the production.

**Reply by sormuras** [reply]
Here's a summary laying out the initial idea of classloader isolation: https://github.com/sormuras/junit-platform-isolator/issues/1#issuecomment-420668480

I am not longer 100% sure how Maven's "provided" scope plays into this picture.

**Reply by zregvart** [reply]
> I am not longer 100% sure how Maven's "provided" scope plays into this picture.

I think the answer to that comes from the answer to the question _Is junit-platform-isolator a container?_, or finding a definition of what _container_ in this case would be.

To me answer to that question would be _yes_, and the dependencies in `provided` scope should be added to the _main_ classloader. The reason I'm saying that is that in production the production code will have access to the classes in `provided` scope, they will be perhaps in a different classloader, in JEE containers that would commonly a parent classloader.

So the _main_ classloader, in my opinion, should have classes from `runtime` and `provided` scopes.

**Reply by TomerFi** [reply]
I second that.

**Reply by sormuras** [reply]
Fine. So, #80 can be merged, I guess.

**Reply by zregvart** [reply]
~@sormuras well, `compile` != `provided` :(~ (not really relevant)

**Reply by zregvart** [reply]
Yeah, though I've looked at the MavenProject source code, the #80 will bring in `compile`, `provided`, `system` and `runtime`.

So the difference from previous state in #77 is the addition of `provided` and `system`, which I think is okay.

Ref. https://github.com/apache/maven/blob/5acdfb665035c5e57c50a9f8198dee53da674df6/maven-core/src/main/java/org/apache/maven/project/MavenProject.java#L346-L371
Ref. https://github.com/apache/maven/blob/5acdfb665035c5e57c50a9f8198dee53da674df6/maven-core/src/main/java/org/apache/maven/project/MavenProject.java#L403-L424

**Reply by sormuras** [reply]
Thanks for the discussion and the investigations, Tomer and Zoran. Very much appreciated!

Let's see what happens next, when #80 is merged.
